### PR TITLE
[REVIEW] Port NVStrings htoi to cudf strings column

### PR DESCRIPTION
**Pull request by davidwendt**
Port NVStrings htoi (hex to integer) conversion to cudf strings column function.

### CODE REVIEW

File: cpp/include/cudf/strings/convert/convert_integers.hpp [hunk 1]
```diff
@@ -69,5 +69,32 @@ std::unique_ptr<column> to_integers( strings_column_view const& strings,
 std::unique_ptr<column> from_integers( column_view const& integers,
                                        rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Returns a new integer numeric column parsing hexidecimal values from the
```
Comment (harrism) [inline cpp/include/cudf/strings/convert/convert_integers.hpp]:
```suggestion
 * @brief Returns a new integer numeric column parsing hexadecimal values from the
```

File: cpp/src/strings/convert/convert_hex.cu [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+ */
+template <typename IntegerType>
+struct hex_to_integer_fn
+{
+    column_device_view const strings_column;
+
+    /**
+     * @brief Converts a single hex string into an integer.
+     *
+     * Non-hexidecimal characters are ignored.
```
Comment (harrism) [inline cpp/src/strings/convert/convert_hex.cu]:
```suggestion
     * Non-hexadecimal characters are ignored.
```

File: cpp/src/strings/convert/convert_hex.cu [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+// This will convert a strings column into any integer column type.
+std::unique_ptr<column> hex_to_integers( strings_column_view const& strings,
+                                         data_type output_type,
+                                         rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                         cudaStream_t stream = 0)
+{
+    size_type strings_count = strings.size();
+    if( strings_count == 0 )
+        return make_numeric_column( output_type, 0 );
```
Comment (harrism) [inline cpp/src/strings/convert/convert_hex.cu]: I think empty_like would be more cudf-idiomatic.

File: cpp/src/strings/convert/convert_hex.cu [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+} // namespace detail
+
+// external API
+std::unique_ptr<column> hex_to_integers( strings_column_view const& strings,
+                                         data_type output_type,
+                                         rmm::mr::device_memory_resource* mr)
+{
+    return detail::hex_to_integers(strings, output_type, mr );
```
Comment (harrism) [inline cpp/src/strings/convert/convert_hex.cu]: Detail API should take a stream.

File: cpp/include/cudf/strings/convert/convert_integers.hpp (hunk 1 above)
Comment (davidwendt) [reply cpp/include/cudf/strings/convert/convert_integers.hpp]: Well, that is just embarrassing. Why did I think it was spelled 'hexidecimal'?

File: cpp/src/strings/convert/convert_hex.cu (hunk 4 above)
Comment (davidwendt) [reply cpp/src/strings/convert/convert_hex.cu]: This is the external API as specified in the comment at line 154.
The detail API called at line 159 is at line 131: https://github.com/rapidsai/cudf/blob/2681125332260782f8e47950f98441e783847199/cpp/src/strings/convert/convert_hex.cu#L134
It has a defaulted stream parameter.

File: cpp/src/strings/convert/convert_hex.cu (hunk 3 above)
Comment (davidwendt) [reply cpp/src/strings/convert/convert_hex.cu]: How about `make_empty_column` instead?
https://github.com/rapidsai/cudf/blob/5a73e75be813e76f750f3ad1416d8cdc54906bbd/cpp/include/cudf/column/column_factories.hpp#L36

File: cpp/src/strings/convert/convert_hex.cu (hunk 4 above)
Comment (harrism) [reply cpp/src/strings/convert/convert_hex.cu]: My mistake again. :)

**Comment by harrism** [comment]
rerun tests

**Comment by davidwendt** [comment]
rerun tests

File: cpp/src/strings/convert/convert_hex.cu [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        int64_t result = 0, base = 1;
+        const char* str = d_str.data();
+        size_type index = d_str.size_bytes();
+        while( index-- > 0 )
+        {
+            char ch = str[index];
+            if( ch >= '0' && ch <= '9' )
+            {
+                result += static_cast<int64_t>(ch-48) * base;
+                base <<= 4; // multiply by 16
```
Comment (jrhemstad) [inline cpp/src/strings/convert/convert_hex.cu]: The compiler is smart enough to do this optimization for you and it's easier to read this way :) 
```suggestion
                base *= 16; // multiply by 16
```

**Pull request merged**